LPS-200030 As an Instance Administrator, I want my User Groups to be automatically un/provisioned using the SCIM protocol

oliverkecskemety opened the pull request:
[LPS-200030](https://liferay.atlassian.net/browse/LPS-200030)

Rebased https://github.com/liferay-appsec/liferay-portal/pull/1473, left out Group member related commits and moved `gw buildREST` commit as last.

liferay-continuous-integration commented:
CI is automatically triggering the following test suites:
- &nbsp;&nbsp;&nbsp;&nbsp;ci:test:**sf**

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 4 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/ad00ebaade0d187f23db8e4f1821185c61393525">ad00ebaade0d187f23db8e4f1821185c61393525</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/liferay-appsec/liferay-portal/tree/LPS-200030-rebase">LPS-200030-rebase</a><br/>Branch GIT ID: <a href="https://github.com/liferay-appsec/liferay-portal/commit/1d95eae97dcfaa4a1977254d7de3228595491a85">1d95eae97dcfaa4a1977254d7de3228595491a85</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-11.liferay.com/job/test-portal-source-format/7601/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-11.liferay.com/userContent/jobs/test-portal-source-format/builds/7601/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-11.liferay.com/job/test-portal-source-format/7601/">test-portal-source-format#7601</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-11.liferay.com/userContent/jobs/test-portal-source-format/builds/7601/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-appsec/liferay-portal/pull/1556">liferay-appsec#1556</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=2470783257">[master] ci:test:sf - liferay-appsec &gt; liferay-appsec - PR#1556 - 2023-12-05[02:43:31]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-11.liferay.com/job/publish-testray-report/4876/">publish-testray-report#4876</a><br/></div></div>

oliverkecskemety commented:
ci:test:sf

liferay-continuous-integration commented:
<h1>Build started.</h1><p>Jenkins is currently running tests in the <code>sf</code> test suite.</p><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a></p><h4>Job Summary:</h4><p>Job Link: <a href="https://test-1-34.liferay.com/job/test-portal-source-format/2976/">test-portal-source-format</a></p><p><b>For more details click <a href="https://test-1-34.liferay.com/userContent/jobs/test-portal-source-format/builds/2976/jenkins-report.html">here</a>.</b></p>

liferay-continuous-integration commented:
<html><h3>:heavy_check_mark: ci:test:sf - 1 out of 1 jobs passed in 3 minutes</h3><details><summary>Click here for more details.</summary><h4>Base Branch:</h4><p>Branch Name: <a href="https://github.com/liferay/liferay-portal/tree/master">master</a><br/>Branch GIT ID: <a href="https://github.com/liferay/liferay-portal/commit/03f7cfaae5c793b488d4684b421ba7c41770d1c0">03f7cfaae5c793b488d4684b421ba7c41770d1c0</a></p><h4>Sender Branch:</h4><p>Branch Name: <a href="https://github.com/liferay-appsec/liferay-portal/tree/LPS-200030-rebase">LPS-200030-rebase</a><br/>Branch GIT ID: <a href="https://github.com/liferay-appsec/liferay-portal/commit/1187efe423767beaa26d2d7862ec76cdfe972ac3">1187efe423767beaa26d2d7862ec76cdfe972ac3</a></p>1 out of 1jobs PASSED<details><summary><strong>1 Successful Jobs:</strong></summary><ul><li><a href="https://test-1-34.liferay.com/job/test-portal-source-format/2976/">test-portal-source-format</a></li></ul></details><h5>For more details click <a href="https://test-1-34.liferay.com/userContent/jobs/test-portal-source-format/builds/2976/jenkins-report.html">here</a>.</h5></details></html>

liferay-continuous-integration commented:
<div><div><strong>Jenkins Build:</strong><span></span><a href="https://test-1-34.liferay.com/job/test-portal-source-format/2976/">test-portal-source-format#2976</a><br/></div><div><strong>Jenkins Report:</strong><span></span><a href="https://test-1-34.liferay.com/userContent/jobs/test-portal-source-format/builds/2976/jenkins-report.html">jenkins-report.html</a><br/></div><div><strong>Jenkins Suite:</strong><span></span>sf<br/></div><div><strong>Pull Request:</strong><span></span><a href="https://github.com/liferay-appsec/liferay-portal/pull/1556">liferay-appsec#1556</a><br/></div><div><strong>Testray Routine:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/builds?testrayRoutineId=1306753894">EE Pull Request</a><br/></div><div><strong>Testray Build:</strong><span></span><a href="https://testray.liferay.com/home/-/testray/runs?testrayBuildId=2471390933">[master] ci:test:sf - liferay-appsec &gt; liferay-appsec - PR#1556 - 2023-12-06[03:42:36]</a><br/></div><div><strong>Testray Importer:</strong><span></span><a href="https://test-1-34.liferay.com/job/publish-testray-report/22019/">publish-testray-report#22019</a><br/></div></div>

oliverkecskemety commented on modules/dxp/apps/scim/scim-rest-impl/src/main/java/com/liferay/scim/rest/internal/manager/UserManagerImpl.java:
```diff
@@ -130,10 +133,7 @@ public User createUser(User user, Map<String, Boolean> requiredAttributes)
-				_getUserGroup(
-					CompanyThreadLocal.getCompanyId(),
-					GetterUtil.getLong(groupId)));
+			_userGroupService.deleteUserGroup(GetterUtil.getLong(groupId));
```
This can throw a `PortalException` on invalid permission

oliverkecskemety commented on modules/dxp/apps/scim/scim-rest-impl/src/main/java/com/liferay/scim/rest/internal/manager/UserManagerImpl.java:
```diff
@@ -787,8 +784,14 @@ private ScimUser _getScimUser(long companyId, long userId)
+			userGroup = _userGroupService.fetchUserGroup(userGroupId);
+		}
+		catch (PortalException portalException) {
+			throw new ForbiddenException(portalException.getMessage());
```
I'm not sure if we want to do it like this but as only the permission check can throw this exception I used a `ForbiddenException` but we can change it if it reveals too much information.

rafaprax commented:
Hey @oliverkecskemety, 

Can we avoid including changes that will be discarded into this pull request(like this [c8d1e0a](https://github.com/liferay-appsec/liferay-portal/pull/1556/commits/c8d1e0a9c43ce64f8d5d4ce22d5ad2cb6d36fa17))?  What do you think?

In my point of view, it keeps the "story" of the PR simpler to follow

Thanks and good job on that.

oliverkecskemety commented on modules/dxp/apps/scim/scim-rest-impl/rest-openapi.yaml:
```diff
@@ -121,6 +136,7 @@ components:
                     description:
                         A multi-valued list of complex objects containing the requested resources.
                     oneOf:
+                        - $ref: "#/components/schemas/GroupSchema"
```
Sadly not. In `BaseGroupResourceTestCase` and `BaseUserResourceTestCase` both `import com.liferay.scim.rest.client.dto.v1_0.Group;` and `import com.liferay.portal.kernel.model.Group;` will exist which causes a collision I cannot fix because even if I use one of them with its fully qualified form the next `gw buildRest` will overwrite it and switch back to the original which breaks the build. If you have any idea I'm open to it.

rafaprax replied:
I see now. Thanks for explaining.

oliverkecskemety commented:
Hey @rafaprax,

Thank you for your review and suggestions. I did what you asked and rebased everything to the current master. Stian usually prefers to have everything in the commit history but I see your point and it might be a bit too much in this case so I left out the unnecessary commits. I changed the access issue handling a bit and followed you advice to refactor the `_getScimClientId` and `_saveScimClientId` functions.

I pushed it to a separate branch so I don't need to force-push and overwrite anything here. Continued in https://github.com/liferay-appsec/liferay-portal/pull/1567
